langchain/src/vectorstores/singlestore.ts

```diff
@@ -0,0 +1,135 @@
+import type {
+  Pool,
+  RowDataPacket,
+  OkPacket,
+  ResultSetHeader,
+  FieldPacket,
+} from "mysql2/promise";
+import { VectorStore } from "./base.js";
+import { Embeddings } from "../embeddings/base.js";
+import { Document } from "../document.js";
+
+export interface SingleStoreVectorStoreConfig {
+  connectionPool: Pool;
+  tableName?: string;
+  contentColumnName?: string;
+  vectorColumnName?: string;
+  metadataColumnName?: string;
+}
+
+export class SingleStoreVectorStore extends VectorStore {
+  connectionPool: Pool;
+
+  tableName: string;
+
+  contentColumnName: string;
+
+  vectorColumnName: string;
+
+  metadataColumnName: string;
+
+  constructor(embeddings: Embeddings, config: SingleStoreVectorStoreConfig) {
+    super(embeddings, config);
+    this.connectionPool = config.connectionPool;
+    this.tableName = config.tableName ?? "embeddings";
+    this.contentColumnName = config.contentColumnName ?? "content";
+    this.vectorColumnName = config.vectorColumnName ?? "vector";
+    this.metadataColumnName = config.metadataColumnName ?? "metadata";
+  }
+
+  async createTableIfNotExists(): Promise<void> {
+    await this.connectionPool
+      .execute(`CREATE TABLE IF NOT EXISTS ${this.tableName} (
+      ${this.contentColumnName} TEXT CHARACTER SET utf8mb4 COLLATE utf8mb4_general_ci,
+      ${this.vectorColumnName} BLOB,
+      ${this.metadataColumnName} JSON);`);
+  }
+
+  async addDocuments(documents: Document[]): Promise<void> {
+    const texts = documents.map(({ pageContent }) => pageContent);
+    const vectors = await this.embeddings.embedDocuments(texts);
+    return this.addVectors(vectors, documents);
+  }
+
+  async addVectors(vectors: number[][], documents: Document[]): Promise<void> {
+    await this.createTableIfNotExists();
+    const { tableName } = this;
+
+    await Promise.all(
+      vectors.map(async (vector, idx) => {
+        try {
+          await this.connectionPool.execute(
+            `INSERT INTO ${tableName} VALUES ('${
+              documents[idx].pageContent
```

Comment:
I don't think we can assume documents are friendly/sanitized - can you use a prepared statement here?

Reply:
It's just good practice regardless